### ISSUE TITLE
Identity Values Not Showing in Checks

### ISSUE THREAD

**Issue by jacobgexigo** [trimble-oss/dba-dash#446]
It seems that the Azure SQL DB databases aren't collecting (or may just not showing) the high value identity columns in the Checks.  We have a database that is in Azure SQL DB for production but in an installed SQL 2019 instance for the dev.  The GUI showed me 99.7% full on a table on dev and I thought there was an issue because prod didn't show up with the same, but I went and looked at the actual table in SSMS and turns out prod was also 99.7% full, I just didn't see it here.  I've checked the DBADash database tables and don't see the data there, so thinking maybe it's the collection of the data.

**Comment by DavidWiseman** [comment]
This should be working for AzureDB.  What version of DBA Dash are you using?

To see if the collection ran:
Navigate to the database in the tree and expand it.  
Click the "Checks" node underneath the database
Click "Collection Dates" tab 
Check the Snapshot Date/Age of the row with reference "IdentityColumns".

You can also check the "DBA Dash ErrorLog" tab to see if there are any errors for this collection.

It's possible to disable collections in the service config tool which is a potential reason why it might not be collected.  ("Configure Schedule" button in the options tab or "Schedule" link on "Source" tab at instance level.)

It's [this query](https://github.com/trimble-oss/dba-dash/blob/main/DBADash/SQL/SQLIdentityColumns.sql) that gets run.  You could try running this against the DB to troubleshoot.  You will need to add this to the beginning of the script:
`DECLARE @IdentityCollectionThreshold INT = 5`

Hope this helps - let me know what you find.

**Comment by jacobgexigo** [comment]
I always keep up to date on the versions (currently on 2.26.0).  

Looks like the snapshot says it is running:
![image](https://user-images.githubusercontent.com/29784949/204017167-ae1a4cff-7267-436a-8835-8a170f6abf7d.png)

There is nothing showing in the Identity Columns for the database:
![image](https://user-images.githubusercontent.com/29784949/204017326-aedbfbbe-9d42-4645-bbe4-e8072ddaf99e.png)

But in the actual database there is this:
![image](https://user-images.githubusercontent.com/29784949/204017777-60327c05-956f-4d5b-ac6b-907e1de3ab21.png)

**Comment by DavidWiseman** [comment]
Can you run the collection query to see what it returns:

```sql
DECLARE @IdentityCollectionThreshold INT = 5
SET NOCOUNT ON 
SET TRAN ISOLATION LEVEL READ UNCOMMITTED

DECLARE @DBName SYSNAME
DECLARE @IdentSQL NVARCHAR(MAX)
DECLARE @SQL NVARCHAR(MAX)
SET @IdentSQL = N'SELECT DB_ID() AS database_id,
	IC.object_id,
	T.name AS object_name,
	IC.name AS column_name,
	CAST(IC.last_value AS BIGINT) AS last_value,
	RC.row_count,
	IC.system_type_id,
	IC.user_type_id,
	IC.max_length,
	CAST(IC.increment_value AS BIGINT) AS increment_value,
	CAST(IC.seed_value AS BIGINT) AS seed_value,
	S.name AS schema_name
FROM sys.identity_columns IC
INNER JOIN sys.tables T ON IC.object_id = T.object_id
INNER JOIN sys.schemas S ON T.schema_id = S.schema_id
OUTER APPLY(SELECT	CASE IC.max_length
						WHEN 1 THEN POWER(2.,IC.max_length*8) 
							ELSE POWER(2.,IC.max_length*8-1)-1 
					END AS max_ident,
					POWER(2.,IC.max_length*8) AS max_rows,
					CAST(IC.last_value AS BIGINT) as last_value_big
			) calc
OUTER APPLY(SELECT SUM(PS.row_count) row_count
			FROM sys.dm_db_partition_stats PS
			WHERE PS.object_id = IC.object_id
			AND PS.index_id < 2 -- HEAP/CLUSTERED
			) RC
WHERE (
	/* last_value is more than threshold percent of the max identity value */
	calc.last_value_big / calc.max_ident * 100 > @IdentityCollectionThreshold 
	/* Table row count is more than the threshold percent of the max number of rows (taking negative values into account)  
	   This is useful if identity was started with a negative number or if the identity was reseeded later with a negative number
	*/
	OR RC.row_count  / calc.max_rows * 100 > @IdentityCollectionThreshold 
	)
AND IC.max_length < 9 /* Exclude decimal types that would be larger than BIGINT and break calculations */
ORDER BY object_name;'

CREATE TABLE #ident(
    database_id SMALLINT NOT NULL,
    object_id INT NOT NULL,
    object_name NVARCHAR(128)  NULL,
    column_name NVARCHAR(128) NULL,
    last_value BIGINT NULL,
    row_count BIGINT NULL,
    system_type_id TINYINT NOT NULL,
    user_type_id INT NOT NULL,
    max_length SMALLINT NOT NULL,
    increment_value BIGINT NULL,
    seed_value BIGINT NULL,
	schema_name NVARCHAR(128) NULL
);


DECLARE DBs CURSOR FAST_FORWARD READ_ONLY LOCAL FOR
			SELECT D.name
			FROM sys.databases D
			WHERE state = 0
			AND HAS_DBACCESS(D.name) = 1
			AND D.is_in_standby = 0
			AND D.database_id <> 2
			AND DATABASEPROPERTYEX(D.name, 'Updateability') = 'READ_WRITE';

OPEN DBs;

WHILE 1=1
BEGIN
	FETCH NEXT FROM DBs INTO @DBName
	IF @@FETCH_STATUS<>0
		BREAK
	IF HAS_DBACCESS(@DBName)=1
	BEGIN
		SET @SQL =  N'USE ' + QUOTENAME(@DBName)  + '
		' + @IdentSQL

		INSERT INTO #ident
		(
		    database_id,
		    object_id,
		    object_name,
		    column_name,
		    last_value,
		    row_count,
		    system_type_id,
		    user_type_id,
		    max_length,
		    increment_value,
		    seed_value,
			schema_name
		)
		EXEC sp_executesql @SQL,N'@IdentityCollectionThreshold INT',@IdentityCollectionThreshold
	END
END
CLOSE DBs 
DEALLOCATE DBs

SELECT database_id,
       object_id,
       object_name,
       column_name,
       last_value,
       row_count,
       system_type_id,
       user_type_id,
       max_length,
       increment_value,
       seed_value,
	   schema_name
FROM #ident

DROP TABLE #ident
```

**Comment by jacobgexigo** [comment]
Yeah, it doesn't come back with any rows, but I think I see why now.  In the prod database the "current identity value" is what we care about in this collection (305713) and the DBCC also returns "current column value" (2145260043), but those aren't necessarily related in the context of the identity itself.  If the value weren't specified on an insert, the next value would be 305714 and therefore is not an issue.  The fact that someone inserted a value so high has nothing to do with the identity unless a RESEED command is issued (then we would have a problem).

In our dev spin up process, we issue RESEED on all tables and that's why it is showing up in the GUI for dev but not prod.  

So long story short, all is well with DBA Dash.  :)

**Comment by DavidWiseman** [comment]
Thanks.  DBCC CHECKIDENT is returning the current identity value of 305,713 which is much less than the threshold of 5% for collection.  The row count of the table must also be under 5%.  

There is a danger in reseeding identities that you could end up with duplicate values or errors if there is a primary key or unique index on the identity column.

:warning: When you ran:
```sql
DBCC CHECKIDENT('xxxxxxx.ContractSummary')
```

I believe this will have caused the identity value to be reseeded to the highest value in the table.   See remarks for `DBCC CHECKIDENT ( table_name )` [here](https://learn.microsoft.com/en-us/sql/t-sql/database-console-commands/dbcc-checkident-transact-sql?view=sql-server-ver16#remarks).

If you want to check the value of the identity without changing it you must use:
```sql
DBCC CHECKIDENT('dbo.IdentityManual',NORESEED)
```
Or you can get the last_value from sys.identity_columns.
